### ISSUE TITLE
Jupyterhub for Kubernes: How to avoid Image Prepull on specific nodes 

### ISSUE THREAD

**Issue by adr-arroyo** [jupyterhub/zero-to-jupyterhub-k8s#1739]
### Bug description
I have a custom Kubernetes cluster with several nodes, each node has assigned some tags to differenciate between them with a Kubernetes label selector (nodeSelector), being one of them PU = "gpu" || "cpu". I have also deployed a Jupyterhub environment on my cluster. However I am having trouble with the deployments of different profiles (singleUser.profileList) on different nodes.

More specifically, I would like to have a profile assigned to nodes with PU: "cpu" and other profile assigned to nodes with PU: "gpu", in order to take advantage of the GPU when deploying notebooks. In the "cpu" profile I have to select the intel version of the  docker image, whereas in the "gpu" profile I need a specific ARM (+ NVIDIA) docker image since the node is deployed on an arm-based machine. I believe I have configured properly each profile so that the user server is deployed in the proper nodes, with the nodeSelector tag (see below). 

The problem comes with the Jupyterhub image prepuller. I have seen that it tries to pull __all the images of every profile in every node__, even though the nodeSelector tag is configured. Then, in the arm/gpu node it tries to prepull the image of the intel/cpu node   , and on this node it tries to pull the arm/gpu image, ending up in an unsuccessfull installation and jupyterhub not being deployed on the cluster.

Is this a bug on the prepuller or a misconfiguration and the prepuller can be configured to avoid this?

#### Expected behaviour
Jupyterhub pulls the proper and configured image when the server is deployed on the node

#### Actual behaviour
Jupyterhub prepulls images on every node, even though their architecture is not the right one.

### How to reproduce

Kubernetes cluster with hybrid ARM and Intel nodes and deploy different user profiles on each node.

### Your personal set up

 - OS:
Linux-Ubuntu 18.04
 - Version:
 jupyterhub 0.9.0
 - Configuration:
Kubernetes deployment.

### My configuration:

```
...

singleuser:
  nodeSelector:
    ARCH: "intel"
    LAYER: "cloud"
  defaultUrl: "/lab"
  image:
    name: easierai/jupyterhub-core
    tag: latest
  profileList:
    - display_name: "environment"
      description: "Environment preconfigured ."
      default: true
      nodeSelector:
        PU: "cpu"
    - display_name: "Minimal Environment"
      description: "Only default libraries installed."
      kubespawner_override:
        image: jupyter/minimal-notebook
        nodeSelector:
          PU: "cpu"
    - display_name: "one NVIDIA"
      description: "ARM, GPU"
      kubespawner_override:
        image: step21/jupyter-minimal-notebook:arm64
        nodeSelector:
          PU: "gpu"
        extraResource:
          limits:
            nvidia.com/gpu: 1

...
```

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jupyterhub-for-kubernes-how-to-avoid-image-prepull-on-specific-nodes/5537/1

**Comment by adr-arroyo** [comment]
Update:

I have worked around this by disabling the prepuller, although it is not my optimal solution:

```
prePuller:
  hook:
    enabled: false
  continuous:
    enabled: false
```

**Comment by loki1978** [comment]
Try this: node_selector instead of nodeSelector .

 - display_name: "(ETL)Minimal environment (Python 3, Jupyterlab)"
      description: "Performance: CPU 2, Memory 6 GB"
      kubespawner_override:
        image: jupyter/minimal-notebook
        cpu_limit: 2
        cpu_guarantee: 1
        mem_limit: '6G'
        mem_guarantee: '4G'
        node_selector:
           eks-node-group: jupyter-system-64g

**Comment by adr-arroyo** [comment]
Thanks @loki1978 

`node_selector` is the tag to be used under `kubespawner_override` and also in the other components as `hub` or `proxy` ? Or this last components use a different tag for deploying on specific nodes?

**Comment by loki1978** [comment]
> Thanks @loki1978
> 
> `node_selector` is the tag to be used under `kubespawner_override` and also in the other components as `hub` or `proxy` ? Or this last components use a different tag for deploying on specific nodes?

Other places it should be nodeSelector.

**Comment by consideRatio** [comment]
> The problem comes with the Jupyterhub image prepuller. I have seen that it tries to pull all the images of every profile in every node, even though the nodeSelector tag is configured.

Ah, in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1818 I've made it possible to opt out of pulling all the images, but you want to pull certain images on certain nodes, based on node_selector, hm....

I'd say this is a bit too hard to accommodate with the image-puller daemonsets in a PR that isn't introducing too much complexity.

Note that the image-puller deamonset pods will limit itself to schedule on the same nodes which the singleuser pods can arrive at as configured with singleuser.nodeSelector and such, but not considering the kubespawner_overrides nested entries.

I'll consider this closed by #1818, not perfect, but it is a clear mitigation of the issue you experience.
